network-area-diagram-viewer/src/main/resources/svg.js

```diff
@@ -0,0 +1,463 @@
+// TODO(Luma) JavaFX seems to have problems drawing a translated foreign object, try to enclose it in a <g> element?
+// TODO(Luma) update on client: find related hidden nodes and update initial position in metadata?
+
+window.addEventListener('load', function() {
+    var svgDiagram = document.getElementsByTagName("svg")[0];
+    if (svgDiagram != null) {
+        makeDraggableSvg(svgDiagram);
+    }
+});
+
+// Draggable SVG Elements from
+// https://www.petercollingridge.co.uk/tutorials/svg/interactive/dragging/
+function makeDraggableSvg(svg) {
+    var selectedElement = false, offset, transform;
+    var diagram = new Diagram(svg);
+
+    svg.addEventListener('mousedown', startDrag);
+    svg.addEventListener('mousemove', drag);
+    svg.addEventListener('mouseup', endDrag);
+    svg.addEventListener('mouseleave', endDrag);
+
+    function getMousePosition(event) {
+        // From screen coordinates to SVG coordinate system
+        var CTM = svg.getScreenCTM();
+        return {
+            x: (event.clientX - CTM.e) / CTM.a,
+            y: (event.clientY - CTM.f) / CTM.d
+        };
+    }
+
+    function startDrag(event) {
+        var draggableElem = diagram.getDraggableFrom(event);
+        if (!draggableElem) {
+            return;
+        }
+        selectedElement = draggableElem;
+        selectedElement.style.cursor = 'grabbing';
+
+        offset = getMousePosition(event);
+        var transforms = getTransformsEnsuringFirstIsTranslation(svg, selectedElement);
+        // Get initial translation amount
+        transform = transforms.getItem(0);
+        offset.x -= transform.matrix.e;
+        offset.y -= transform.matrix.f;
+    }
+
+    const CONTINUOUS_UPDATE = true;
+    function drag(event) {
+        if (selectedElement) {
+            event.preventDefault();
+            var position = getMousePosition(event);
+            var translation = {x: position.x - offset.x, y: position.y - offset.y};
+            transform.setTranslate(translation.x, translation.y);
+            if (CONTINUOUS_UPDATE) {
+                diagram.update(selectedElement, position, translation);
+            }
+        }
+    }
+
+    function endDrag(event) {
+        if (selectedElement) {
+            console.log("endDrag for element " + selectedElement.id);
+            selectedElement.style.cursor = 'grab';
+            // Desired position for the selected element is current mouse position in SVG coordinate system
+            var position = getMousePosition(event);
+            var translation = {x: position.x - offset.x, y: position.y - offset.y};
+            diagram.update(selectedElement, position, translation);
+        }
+        selectedElement = null;
+    }
+}
+
+// PowSyBl Diagram
+
+function Diagram(svg) {
+    this.getDraggableFrom = getDraggableFrom;
+    this.update = update;
+
+    function getDraggableFrom(event) {
+        if (isDraggable(event.target)) {
```

Comment:
and then here we don't need any `isDraggable`, we only need to get the parentElement until we find a `nad-vl-node` (or until we get `nad-vl-nodes` then we stop). Indeed we don't want to move bus nodes, but their parent element.

Note that we need to add the class `nad-vl-node` in powsybl-diagram.

Reply:
Yes, that will simplify the logic. Great.

Reply:
I tried your proposal, but I think code will be more clear if we try to keep the makeDraggable functions separated from the code that "knows things" about the contents of the diagram.

Also, checking if a node has class vl-node (or boundary-node, or 3wt-node) is a check that answers the question "should this element be considered draggable?", so I think that we must keep the `getDraggableFrom` and `isDraggable` methods, to be clear about what we allow to drag and not, and easily extend it.

In any case, as you mention, there is a simplification over `getDraggableFrom` and `isDraggable`: we have to capture the whole `vl-node `element, not the `busnode`, and navigation through parent can be simplified. Take note of that.

Reply:
I've simplified the logic for checking draggable elements.
We could discuss if we need an explicit class `nad-vl-node` for every drawing corresponding to a node or it could be enough to consider that all direct children of a group `nad-vl-nodes` (this resulting in less verbose SVG). I've implemented both alternatives in current javascript, but used the last one, that does not require additional classes in the SVG.